Local ARS HLD

VladimirKuk opened the pull request:
initial version

VladimirKuk committed:
initial HLD version

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+**Existing Forwarding Decision Model**
+
+Today, the routing protocol or SDN controller decides the reachability of a given destination and finds all possible paths. These paths may be equal cost or unequal cost. Decision to choose one of the paths from all available paths is done in the switch pipeline based on a computed hash. This path selection is static in nature based on the packet fields. Hash based selection doesn't take into acc ...
```
**The** decision to choose of one the paths...

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+Today, the routing protocol or SDN controller decides the reachability of a given destination and finds all possible paths. These paths may be equal cost or unequal cost. Decision to choose one of the paths from all available paths is done in the switch pipeline based on a computed hash. This path selection is static in nature based on the packet fields. Hash based selection doesn't take into acc ...
+
+Control plane protocols exist for traffic engineering a path but involves a control plane decision that is very slow to react to changing traffic patterns in the network or state of interfaces.
```
traffic engineering paths, but involves control plane decisions that are very slow...

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+Control plane protocols exist for traffic engineering a path but involves a control plane decision that is very slow to react to changing traffic patterns in the network or state of interfaces.
+
+Adaptive Routing System (ARS) allows dynamically selection of the best available path for data packets based on real-time network conditions. This approach helps to mitigate congestion, optimize resource utilization, and improve overall network performance.
```
allows dynamic selections of the best...

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+Adaptive Routing System (ARS) allows dynamically selection of the best available path for data packets based on real-time network conditions. This approach helps to mitigate congestion, optimize resource utilization, and improve overall network performance.
+
+No standard exists, but there are industry consesus for general approach.
```
is an industry **consensus**

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+No standard exists, but there are industry consesus for general approach.
+
+Following illustrates common operation of ARS, in which two paths are considered - one shortest (minimal) and another (non-minimal), when forwarding trafficto the destination. When shortest path becomes congested, alternative will be used.
```
- **The** following illustrates **the** common operation of ARS
- traffic to
- **the** alternative

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+The diagram illustrates a Conceptual Packet Flow, where macro flows are collections of multiple micro flows that share the same routing destination. Micro flows, identified by a 5-tuple (source IP, destination IP, protocol, source port, destination port), are hashed, and if several micro flows hash to the same bucket, they are grouped into a macro flow. These flows are further segmented into flow ...
+
+__Figure 2: ARS Packet Flow__
+![ARS flow](images/ARSPacketFlow.png "Figure 1: ARS Packet Flow")
```
Is it possible to put a white background on the image? opening it in dark mode makes it hard to read

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+This diagram outlines a Logical Pipeline Flow for Adaptive Routing with ARS. Flow starts with L3 Route Lookup to identify the appropriate Next-Hop Group (NHG). If ARS is enabled, the system generates an ARS Macro Flow ID and assigns packets to specific next-hop destinations using a Macro Flow Table. Otherwise, traffic undergoes standard ECMP processing. In the background, ARS monitors port load m ...
+
+__Figure 3: ARS SAI Pipeline Flow__
+![SAI model](images/ARSPipeline.png "Figure 2: ARS Pipeline Flow")
```
Also this give a white background?

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+arsorch - it is an orchestration agent that handles the configuration requests from CONFIG_DB or APPL_DB. It is responsible for creating the SAI ARS profile and configuring SAI ARS object.
+
+#### Orchdaemon
+orchdaemon - it is the main orchestration agent, which handles all Redis DB's updates, then calls appropriate orchagent, the new arsorch should be registered inside an orchdaemon.
```
**The** orchdaemon

amazor commented on doc/ARS/Local_ARS_HLD.md:
````diff
@@ -0,0 +1,732 @@
+##### ARS_PORIFLE
+
+```
+	container ARS_PORIFLE {
````
ARS_PROFILE

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+	container ARS_OBJECT {
+
+		list ARS_OBJECT_LIST {
+			key "profile_name ars_name";
```
two keys?

amazor commented on doc/ARS/Local_ARS_HLD.md:
````diff
@@ -0,0 +1,732 @@
+##### ACL_RULE changes
+
+```
+	container sonic-acl {
````
remove all here except for the new leaf

amazor commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+						length 1..255;
+					}
+				} 
+...
```
remove extra "```"

DanielaMurin commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+| Revision | Date        | Author           | Change Description |
+| -------- | ----------- | ---------------- | ------------------ |
+| 1.0      | Dec 01 2024 | Vladimir Kuk     | Initial proposal   |
```
Please add Marvell somewhere to the revision.

DanielaMurin commented on doc/ARS/Local_ARS_HLD.md:
````diff
@@ -0,0 +1,732 @@
+	}
+```
+
+#### Config DB Enhancements  
````
Please fix formatting issues in "Config DB Enhancements"

DanielaMurin commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+### Warmboot and Fastboot Design Impact  
+
+Warmboot for SWSS container will be supported.
```
If there is no delay or impact on warmboot, that should be mentioned here.

DanielaMurin commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,732 @@
+
+### Restrictions/Limitations
+
+Due to hw resources allocations, all ARS configuration must be done prior to routing configuration.???????????
```
Let's remove the ???????????
If this is an issue, let's move it to the "Open/Action items" section.

VladimirKuk committed:
review changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Identation fix

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Identation fix

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
changes after initial review

VladimirKuk committed:
update fonts

VladimirKuk committed:
intergrated ARS_OBJECT in NHG and LAG, use notification mechanism

VladimirKuk committed:
removed ARS_OBJECT

VladimirKuk committed:
Update to latest changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Review changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+This design follows SAI conceptual model described in https://github.com/opencomputeproject/SAI/blob/master/doc/ARS/Adaptive-Routing-and-Switching.md.
+
+
+The diagram illustrates a Conceptual Packet Flow, where macro flows are collections of multiple micro flows that share the same routing destination. Micro flows, identified by a 5-tuple (source IP, destination IP, protocol, source port, destination port), are hashed, and if several micro flows hash to the same bucket, they are grouped into a macro flow. These flows are further segmented into flow ...
```
he diagram illustrates a Conceptual Packet Flow, where macro flows are collections of multiple micro flows ~~that share the same routing destination~~. Micro flows, identified by a 5-tuple (source IP, destination IP, protocol, source port, destination port), are hashed, and if several micro flows hash to the same bucket, they are grouped into a macro flow. These flows are further segmented into flow-lets based on idle time thresholds between packets. A Macro Flow Table maps these flows to next-hop destinations and tracks their status (e.g., "active" or "expired"). Using Adaptive Routing within an ECMP group, the system dynamically assigns traffic based on link quality metrics like latency and packet loss.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+
+### Use cases
+- L3 traffic egressing via NHG
+    * NH pointing to portlag
```
portlag is a typo? Did u mean port.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+Configuration exmaple:
+
+"ARS_INTERFACE": {
+    "flowlet-based | Ethernet0" : {},
```
Does not align with the key defined earlier, and flowlet mode is an ARS/NHG property and not that of port.

```
"ARS_INTERFACE": {
    “Ethernet0" : {
               enable: true,
               scaling-factor: 100 // if user wants to override the scaling factor derived based on port speed
    }
} 
```

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+        "profile_name": "ars_profile",
+        "assign_mode" : "per_flowlet_quality",
+        "flowlet_idle_time" : 256,
+        "max_flows" : 512,
```
1. Edit of this attribute is not supported in TL s/w.
2. Supported values on TL10 are: [32, 64, 128, 256, 512, 1024, 2048, 4096]

How do we ensure that we do not send unsupported operations/values to TL SAI? 
- Is it ok to round off to powers of 2 for all platforms?
- We need to avoid a set operation on TL SAI

VladimirKuk replied:
It will be hard to force this behavior on all vendors. Much easier to explain to the customer what are the valid values or maybe round in TL SAI.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+    "PortChannel1" : {
+        "profile_name": "ars_profile",
+        "assign_mode" : "per_flowlet_quality",
+        "flowlet_idle_time" : 256,
```
TL SAI support is upto 2047 usecs. Is it possible to cap it in SONiC?

VladimirKuk replied:
I can try adding to yang, but if some vendor will try changing it later on, it still could be a problem.
What is the minimum value, it is 2 or 1 ?

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+
+Tests separated into two group - mandatory and optional (only if supported by vendor sai) parts.
+
+- Manadatory:
```
typo

VladimirKuk replied:
thanks

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+ipv6_enable             = boolean       ;Whether ARS is enabled over IPv6 packets
+
+
+Configuration exmaple:
```
pls fix typo here and everywhere else.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+
+profile_name             = string                                   ;ARS profile Name
+assign_mode              = "per_flowlet_quality" / "per_packet"     ;member selection assignment mode
+flowlet_idle_time        = uint16                                   ;idle time for decting flowlet in macro flow. Relevant only for assign_mode=pre_flowlet
```
Max value supported in TL SAI is 2047 usecs

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+* Phase 1
+
+| SAI api | Supported SAI attribute 
+| ------- | ----------------------- 
```
Add CRM support using SAI_ARS_PROFILE_ATTR_ECMP_ARS_MAX_GROUPS/MAX_MEMBERS_PER_GROUP

VladimirKuk replied:
These are READ_ONLY attributes. How do you propose to use them ? Validate number of nexthop groups or number of nexthops  before configuration ?

rck-innovium replied:
please take a look at [orchagent/crmorch.cpp](https://github.com/sonic-net/sonic-swss/pull/1974/files#diff-c3579ec2f250e54d61b7b9a7ee39aedef555d465970cc0e613ab01efea7ff22a)

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+    * Validate behavior when enabling ARS on unsupported interfaces.
+
+3. ARS for Nexthop Groups (L3 traffic)
+    * Confirm that ARS correctly applies to nexthop groups for Adaptive Routing.
```
During test implementation, ensure that the traffic expectation is that the tx port has only one ARS NH on it.

VladimirKuk replied:
Ok, no need to mention in HLD ?

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+
+"ARS_INTERFACE": {
+    "flowlet-based | Ethernet0" : {},
+    "flowlet-based | Ethernet8" : {}
```
 It would be good if we can expand the design section to cover the sequence when user enables ARS on an already created port. I am assuming based on the above config, Orchagent will set:
```
SAI_PORT_ATTR_ARS_ENABLE= True,
SAI_PORT_ATTR_ARS_PORT_LOAD_PAST_ENABLE=True
SAI_PORT_ATTR_ARS_PORT_LOAD_SCALING_FACTOR= value given? Value : Port Speed div by 1G
```

VladimirKuk replied:
I'm actually covering this sequence in code but forgot to reflect in HLD, thanks.
Do you see non-default port scaling use-case ?

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+algorithm               = "ewma"        ;Path quality calculation algorithm
+max_flows               = uint32        ;Maximum number of flows that can be maintained for ARS
+sampling_interval       = uint32        ;Sampling interval in microseconds
+past_load_min_value     = uint16        ;Minimum value of Past load range.
```
This needs to be 0 on Teralynx. Do you see any need to have past_load_min_value as a value other than 0.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+future_load_weight      = uint16        ;Weight of the future load
+current_load_min_value  = uint16        ;Minimum value of Current load range.
+current_load_max_value  = uint16        ;Maximum value of Current load range.
+ipv4_enable             = boolean       ;Whether ARS is enabled over IPv4 packets
```
both of these (ipv4/ipv6 enable) are by default turned on by TL SAI and cannot be changed to false.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+
+2. Set Port Scaling Factor
+Based on the port's speed, PortsOrch determines the appropriate scaling factor.
+The scaling factor is then set in the SAI layer through the syncd component to adjust the port's behavior accordingly.
```
Does SONiC always give a link up notification when the port speed is changed?

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+assign_mode              = "per_flowlet_quality" / "per_packet"     ;member selection assignment mode
+flowlet_idle_time        = uint16                                   ;idle time for decting flowlet in macro flow. Relevant only for assign_mode=pre_flowlet
+max_flows                = uint16                                   ;Max number of flows supported for ARS
+primary_path_threshold   = uint16                                   ;Quality threshold for primary path 
```
Primary/Alternate is not supported in TL S/W and AFAIK there is no special support in the device for these. If really needed, we need to manage in s/w.

rck-innovium commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,860 @@
+    * If LAG was created:
+        - It is updated with the ARS object ID.
+
+__Figure 10:LAG creation flow__
```
Consider moving the support for LAG (SAI_LAG_ATTR_ARS_OBJECT_ID) to phase 2. This is supported by IFCS but not in TL SAI since it was not needed by customers so far.

rck-innovium reviewed (commented):
We need to devise a strategy on how to ensure that we do not pass unsupported operations/attributes/values to TL SAI. Ensure that we have capability queries where ever possible, and round-offs in other places.

VladimirKuk committed:
Addressing comments

Fixed phases contents
Fixed spelling
Added user configured port scaling factor
Updated init sequence
Limit idle time value

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Addressing comments

Fixed phases contents
Fixed spelling
Added user configured port scaling factor
Updated init sequence
Limit idle time value

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

Gnanapriya27 commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+4. Support ACL action to disable ARS 
+
+* Phase 2
+    - Support enabling ARS over LAG
```
Hope HLD covers ARS over ECMP and LAG and the implementation is planned to be phased out manner , starts with ECMP first and then the LAG . HLD requirement section should reflect the HLD status i believe.

Gnanapriya27 commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+#### Sequence diagrams
+
+__Figure 4: Initilization flow__
+![](images/init_seq.png)
```
"Bind to Event" - which is very generic, can you callout the list of port events that are consumed by ARSorch.

Gnanapriya27 commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+__Figure 7: Nexthop group creation flow__
+![](images/nhg_create.png)
+
+1. Nexthop group creation
```
NextHop Group Deletion flow is not covered in this HLD, which is essential for the completion of the SONiC HLD.

Gnanapriya27 commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+__Figure 8: Nexthop group update flow__
+![](images/nhg_update.png)
+
+1. Nexthop group update
```
Update flow covers the group member update and didnt cover the update flow from ARS enable to disable or vice versa - please check

Gnanapriya27 commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+
+**Existing Forwarding Decision Model**
+
+Today, the routing protocol or SDN controller decides the reachability of a given destination and finds all possible paths. These paths may be equal cost or unequal cost. The decision to choose of one the paths from all available paths is done in the switch pipeline based on a computed hash. This path selection is static in nature based on the packet fields. Hash based selection doesn't take into ...
```
First para alone i attempted to modify in order to emphasize the problem this HLD is trying to address , please check i if it make sense, this can be used- "Today, routing protocols or SDN controllers establish destination reachability and compute all possible paths, whether equal-cost or unequal-cost. However, the selection of a specific path from the available ECMP paths is performed statically in the switch data plane, driven by a hash computed over packet header fields. This static hashing mechanism does not account for real-time path conditions of the local or end-to-end path, as a result, once a path is chosen, the packet flow remains locked to that path, leading to inefficient load balancing and potential network hotspots."

Gnanapriya27 reviewed (commented):
Have reviewed and shared feedback

VladimirKuk committed:
Addressing community comments

Separated NHG identification and ARS object
Added NHG matching by ARS-enabled members
Support NH creation from RouteOrch and NhgOrch

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Fixing Review comments

Signed-off-by: Ashok Kumar P <apannerselva@marvell.com>
